### PR TITLE
[#196] Implement incremental write loop for write operation

### PR DESCRIPTION
**Pull request by korydraughn**
All tests pass.

Considering the idea of passing an input struct rather than 10+ parameters. Not required, but much easier to deal with from a code standpoint and while reading. Can also handle this later.

TODO: Add tests which force the use of bg thread pool when writing to iRODS. The number of writes depend on the configuration of the HTTP API, so we'll need to document the assumptions of the test suite too.

### CODE REVIEW

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [inline endpoints/data_objects/src/main.cpp:450]: Document these

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [inline endpoints/data_objects/src/main.cpp:374]: Consider passing via an input struct.

**Comment by korydraughn** [comment]
Added tests.

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [inline endpoints/data_objects/src/main.cpp:428]: Handle TODO

File: endpoints/data_objects/src/main.cpp
Comment (alanking) [inline endpoints/data_objects/src/main.cpp:417]: So rather than having a single foreground task which loops to stream bytes, we've switched to a recursive background task which spawns background tasks from itself for each write (itself also being a background task) until there's nothing left to write. Is that right?

If so, why the switch to this model? Just trying to understand.

File: test/test_irods_http_api.py
Comment (alanking) [inline test/test_irods_http_api.py:1655]: If it's possible for the HTTP API to check its own configuration, we can skip or fail the test if it's not meeting the assumptions. Just a thought; don't have to do that.

File: endpoints/data_objects/src/main.cpp
Comment (alanking) [inline endpoints/data_objects/src/main.cpp:391]: Maybe add `// constructor` here?

File: test/test_irods_http_api.py
Comment (alanking) [inline test/test_irods_http_api.py:1681]: Should we only set this to `True` after `parallel_write_init` has completed successfully? I think `parallel_write_handle` will not be defined should either of the assumptions following `parallel_write_init` fail, which would cause problems in the cleanup. Until `parallel_write_init` succeeds, I think it should be set to `False`.

Also, I would argue the word "primarily" can be removed from the comment as it doesn't appear to be used anywhere else besides the cleanup.

**Review by alanking — COMMENTED** [review]
Seems good overall. Had a couple questions, couple suggestions

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [reply endpoints/data_objects/src/main.cpp:417]: The switch was inevitable. If we don't break up long running tasks, then the server runs the risk of becoming unresponsive. There are a limited number of threads and they have to devote time to each request. If too many clients appear and each of them attempt to upload a file which causes a thread to have to enter a loop, clients will feel the latency.

File: test/test_irods_http_api.py
Comment (korydraughn) [reply test/test_irods_http_api.py:1655]: The tests do not have that ability right now. We can add that though.

That information could be exposed via the /info endpoint. The question then becomes, is that information valuable to users or is it only valuable to the tests.

Please open an issue for that.

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [reply endpoints/data_objects/src/main.cpp:417]: Sure thing.

File: test/test_irods_http_api.py
Comment (alanking) [reply test/test_irods_http_api.py:1655]: Aren't these tests running on the same machine as the HTTP API server? If so, we could just crack open the config file, right? Sorry if these are naive questions.

Comment (korydraughn) [reply test/test_irods_http_api.py:1655]: It's fine. Ask all the questions.

> Aren't these tests running on the same machine as the HTTP API server?

They aren't required to run on the same machine. The config.py provides information for helping the tests run against local or remote HTTP API servers. The tests do not assume the HTTP API is local to them.

I run the tests locally, but others have been known to run/test it remotely.

Comment (alanking) [reply test/test_irods_http_api.py:1655]: Okay, that makes sense. Maybe we can set the configurations in question in the test config and leave a comment referencing this test should anybody ever change these? Just concerned about the fact that these tests can (and will?) pass even if the assumptions are not met.

Anyway, I created https://github.com/irods/irods_client_http_api/issues/322 for the configuration fetching mechanism.

Comment (korydraughn) [reply test/test_irods_http_api.py:1655]: That's a possibility. The GenQuery2 tests are controlled by config.py just like you've described.

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [reply endpoints/data_objects/src/main.cpp:374]: We can revisit this at a later time. The current implementation works and it's unlikely that anyone will need to modify this chunk of code once merged.

I'll open an issue to revisit this at a later time.

File: test/test_irods_http_api.py
Comment (korydraughn) [reply test/test_irods_http_api.py:1655]: Resolving now that we have issue #322.

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [reply endpoints/data_objects/src/main.cpp]: Done

File: test/test_irods_http_api.py
Comment (korydraughn) [reply test/test_irods_http_api.py]: Done

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [reply endpoints/data_objects/src/main.cpp]: Done.

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [reply endpoints/data_objects/src/main.cpp]: Done.

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [reply endpoints/data_objects/src/main.cpp]: Done.

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [reply endpoints/data_objects/src/main.cpp]: Created #326.

**Comment by korydraughn** [comment]
All review comments have been resolved.

Tests still pass.

Waiting for final review.

File: test/test_irods_http_api.py
Comment (korydraughn) [inline test/test_irods_http_api.py:1704]: 3 -> `stream_count`

Comment (korydraughn) [reply test/test_irods_http_api.py]: Done

**Comment by korydraughn** [comment]
Made minor tweak to test - no change in behavior.

Tests still pass.

Ready for final review again.

**Review by alanking — APPROVED** [review]
I'd feel better with one more set of eyeballs before #s. Up to you

**Comment by trel** [comment]
looks good to me - i think ready for squashing.

**Review by trel — APPROVED** [review]
yep, confirmed it looks the same.   separation is good on the two commits.    

wait for one more before #ing

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [inline endpoints/data_objects/src/main.cpp:372]: Could make it more obvious that ownership of some inputs is being transferred by requiring an R-value.

Making that change would only affect the call site. There would be no change in behavior.

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [inline endpoints/data_objects/src/main.cpp:949]: Consider renaming `/irods_client/buffer_size_in_bytes_for_write_operations` config option.

It is not a buffer size. It is the max number of bytes to write to iRODS in a single call.

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [reply endpoints/data_objects/src/main.cpp]: Done.

File: endpoints/data_objects/src/main.cpp
Comment (korydraughn) [reply endpoints/data_objects/src/main.cpp]: Renamed to `max_number_of_bytes_per_write_operation`.

Uses the same naming scheme as the option for read operations.

**Comment by korydraughn** [comment]
Tests still pass.

The latest commits do not introduce any changes in behavior.

**Comment by trel** [comment]
good new name.   looks good.

**Review by alanking — COMMENTED** [review]
Squash em, I say

**Comment by korydraughn** [comment]
Squashed.

@MartinFlores751's review is on the way.

**Review by MartinFlores751 — APPROVED** [review]
Looks good to me.

To note, Kory and I did talk yesterday about the 'boost idiom' of the 'recursion with a shared pointer to self' just to understand why it was used vs putting something similar to the previous code in a background task instead. Similar question: https://github.com/irods/irods_client_http_api/pull/320#discussion_r1725022659

**Review by alanking — APPROVED** [review]
\# it if happy with the changes.

**Pull request merged**
